rosbag2_transport/src/rosbag2_transport/player.cpp

```diff
@@ -357,27 +359,44 @@ rosbag2_storage::SerializedBagMessageSharedPtr * Player::peek_next_message_from_
   return message_ptr;
 }
 
-bool Player::play_next()
+bool Player::play_next(const std::optional<uint64_t> num_messages)
 {
+  // Evaluates the escape condition in which no message is required to be published.
+  if (num_messages.has_value() && *num_messages == 0) {
+    return true;
+  }
+
   // Temporary take over playback from play_messages_from_queue()
   std::lock_guard<std::mutex> lk(skip_message_in_main_play_loop_mutex_);
 
   if (!clock_->is_paused() || !is_in_play_) {
     return false;
   }
 
-  skip_message_in_main_play_loop_ = true;
-  rosbag2_storage::SerializedBagMessageSharedPtr * message_ptr = peek_next_message_from_queue();
-
-  bool next_message_published = false;
-  while (message_ptr != nullptr && !next_message_published) {
-    {
-      rosbag2_storage::SerializedBagMessageSharedPtr message = *message_ptr;
-      next_message_published = publish_message(message);
-      clock_->jump(message->time_stamp);
+  bool next_message_published{false};
+  uint64_t message_countdown = num_messages.has_value() ? *num_messages : 1u;
+
+  while (message_countdown > 0) {
+    // Resets state for the new iteration.
+    next_message_published = false;
+    skip_message_in_main_play_loop_ = true;
+    // Grabs a message and tries to publish it. When there is no publisher with
+    // the message->topic_name, it just peeks the next one.
+    rosbag2_storage::SerializedBagMessageSharedPtr * message_ptr = peek_next_message_from_queue();
+    while (message_ptr != nullptr && !next_message_published) {
+      {
+        rosbag2_storage::SerializedBagMessageSharedPtr message = *message_ptr;
+        next_message_published = publish_message(message);
+        clock_->jump(message->time_stamp);
+      }
+      message_queue_.pop();
+      message_ptr = peek_next_message_from_queue();
+    }
+    // There are no more messages of the built publishers to play.
+    if (!next_message_published) {
+      break;
     }
-    message_queue_.pop();
-    message_ptr = peek_next_message_from_queue();
+    message_countdown--;
```

Comment:
Should this be inside the inner `while` loop?

Reply:
I've added a docstring in 0381e31 . `message_countdown` should be decreased when a message is effectively published. `publish_message()` already cares of sending the message iff there is matching publisher for its topic. When there is none, the return value is `false` and the previous while loop will continue if there are more messages in the storage. On the contrary, when the message is sent, the flag becomes true and the while loop condition is no longer valid which also makes the next if clause to be skipped and the counter to be decreased.

I hope that docstring helps with the understanding of the code without needing to go back and forth in the file (I had to do it when inspecting similar loops in other functions).